app/src/main/java/net/christianbeier/droidvnc_ng/InputService.java

```diff
@@ -879,4 +874,112 @@ private static void setCursorPos(AccessibilityNodeInfo node, int cursorPos) {
 		action.putInt(AccessibilityNodeInfo.ACTION_ARGUMENT_SELECTION_END_INT, cursorPos);
 		node.performAction(AccessibilityNodeInfo.AccessibilityAction.ACTION_SET_SELECTION.getId(), action);
 	}
+
+	private interface GestureHandler {
+		void startGesture(int x, int y);
+		void continueGesture(InputContext inputContext, int x, int y);
+		void endGesture(InputContext inputContext, int x, int y);
+	}
+
+	private static class LegacyGestureHandler implements GestureHandler {
```

Comment:
no need for these classes IMO, just do the Android version differentiation in the existing handlers .-)

Reply:
What did you had in mind exactly? I'm trying to merge the two handlers, but the code is getting a lot more confusing and hard to follow.

Unrelated: I noticed that you were using `System.currentTimeMillis()`, which is not monotonic. This means that if the system time changes while there is an ongoing gesture the resulting `duration` will be wrong. `SystemClock.elapsedRealtime()` is better for measuring intervals.